pricegraph/src/orderbook.rs

```diff
@@ -146,53 +146,93 @@ impl Orderbook {
+
+        // NOTE: In the case where the market `quote` and `base` token are in
+        // different disconnected subgraphs, it is possible that the `base`
+        // token's subgraph contains negative cycles. In order to ensure that
+        // the `base` token subgraph is also reduced, fill any remaining
+        // negative cycles in the inverse market. However, there should be no
+        // ring trades over the inverse market (if there were, then the `quote`
+        // and `base` token would be part of the same subgraph), so assert it.
+        let inverse_ring = self.fill_market_ring_trade(Market {
+            base: market.quote,
+            quote: market.base,
+        });
+        debug_assert_eq!(inverse_ring, None);
+
+        overlap
+    }
+
+    /// Fills a ring trade over the specified market, and returns the flow
+    /// corresponding to both ask and bid segments of the ring. Returns `None`
+    /// if there are no overlapping ring trades over the specified market.
+    ///
+    /// Note that if this method returns `None`, then the orderbook is
+    /// **partially** reduced. Specifically, the subgraph containing the market
+    /// `quote` token is fully reduced, however, other not connected subgraphs,
+    /// specifically the market `base`'s subgraph in the case where the `quote`
+    /// and `base` token are not part of the same subgraph, may still contain
+    /// negative cycles.
+    pub fn fill_market_ring_trade(&mut self, market: Market) -> Option<Ring> {
         if !self.is_token_pair_valid(market.bid_pair()) {
-            return Default::default();
+            return None;
         }
 
         let (base, quote) = (node_index(market.base), node_index(market.quote));
 
-        let mut overlap = TransitiveOrderbook::default();
         while let Err(NegativeCycle(predecessors, node)) =
             bellman_ford::search(&self.projection, quote)
         {
             let path = path::find_cycle(&predecessors, node, Some(quote))
                 .expect("negative cycle not found after being detected");
 
-            if path.first() == Some(&quote) {
-                if let Some(base_index) = path.iter().position(|node| *node == base) {
-                    let (ask, bid) = (&path[0..base_index + 1], &path[base_index..]);
-
-                    let ask = self
-                        .fill_path(ask)
-                        .expect("ask transitive path not found after being detected");
-                    overlap.asks.push(ask.as_transitive_order());
-
-                    let bid = self
-                        .fill_path(bid)
-                        .expect("bid transitive path not found after being detected");
-                    overlap.bids.push(bid.as_transitive_order());
+            let base_index = match if path.first() == Some(&quote) {
+                path.iter().position(|node| *node == base)
+            } else {
+                None
+            } {
+                Some(value) => value,
```

Comment:
ubernit: I found the two consecutive match statements slightly hard to read. Maybe splitting them into two is a bit easier on the eye?

Reply:
Too bad cause I thought this pattern looked kind of neat :smile: and I read it like "match, if the first node is the quote, then the position of the base token, else match nothing".

That being said, I completely agree it's a little hard to parse at a quick glance and not a pattern we are used to seeing.